File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.py
def solution(numbers):
    answer = [-1]
    stack = [numbers[-1]]
    
    for i in range(len(numbers) - 2, -1, -1):
        while stack:
            if stack[-1] <= numbers[i]:
                stack.pop()     
            else:
                answer.append(stack[-1])
                break
        
        if not len(stack):
            answer.append(-1)
            
        stack.append(numbers[i])
        
    return answer[::-1]

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.*;

class Solution {
    public int[] solution(int[] numbers) {
        int n = numbers.length;
        int[] answer = new int[n];
        Arrays.fill(answer, -1);
        
        Deque<Integer> stack = new ArrayDeque<>();
        
        // 맨 뒤부터 순회
        for (int i = n - 1; i >= 0; i--) {
            int cur = numbers[i];
            
            while (!stack.isEmpty() && stack.peek() <= cur) stack.pop();
            
            if (!stack.isEmpty()) answer[i] = stack.peek();
            // else answer[i] = -1;
            
            stack.push(cur);
        }
        
        return answer;
    }
}